core/toolbox/toolbox.js

```diff
@@ -0,0 +1,459 @@
+ * @return {!Element} The html container for the toolbox contents.
+ * @protected
+ */
+Blockly.Toolbox.prototype.createContentsContainer_ = function() {
+  var contentsContainer = document.createElement('div');
+  Blockly.utils.dom.addClass(contentsContainer, 'blocklyToolboxContents');
+  if (this.isHorizontal()) {
+    contentsContainer.style.flexDirection = 'row';
+  }
+  return contentsContainer;
+};
+
+/**
+ * Adds event listeners to the toolbox container div.
+ * @param {!Element} container The html container for the toolbox.
+ * @param {!Element} contentsContainer The html container for the contents
+ *     of the toolbox.
+ * @protected
+ */
+Blockly.Toolbox.prototype.addToolboxListeners_ = function(container,
+    contentsContainer) {
+  // Clicking on toolbox closes popups.
+  var clickEvent = Blockly.bindEventWithChecks_(container, 'mousedown', this,
+      this.onClick_, /* opt_noCaptureIdentifier */ false,
+      /* opt_noPreventDefault */ true);
+  this.boundEvents_.push(clickEvent);
+
+  var keyDownEvent = Blockly.bindEventWithChecks_(contentsContainer, 'keydown',
+      this, this.onKeyDown_, /* opt_noCaptureIdentifier */ false,
+      /* opt_noPreventDefault */ true);
+  this.boundEvents_.push(keyDownEvent);
+};
+
+/**
+ * Creates the flyout based on the toolbox layout.
+ * @return {!Blockly.Flyout} The flyout for the toolbox.
+ * @throws {Error} If missing a require for `Blockly.HorizontalFlyout`,
+ *     `Blockly.VerticalFlyout`, and no flyout plugin is specified.
+ * @protected
+ */
+Blockly.Toolbox.prototype.createFlyout_ = function() {
+  var workspace = this.workspace_;
+  var workspaceOptions = new Blockly.Options(
+      /** @type {!Blockly.BlocklyOptions} */
+      ({
+        'parentWorkspace': workspace,
+        'rtl': workspace.RTL,
+        'oneBasedIndex': workspace.options.oneBasedIndex,
+        'horizontalLayout': workspace.horizontalLayout,
+        'renderer': workspace.options.renderer,
+        'rendererOverrides': workspace.options.rendererOverrides
+      }));
+  workspaceOptions.toolboxPosition = workspace.options.toolboxPosition;
+
+  var FlyoutClass = null;
+  if (workspace.horizontalLayout) {
+    FlyoutClass = Blockly.registry.getClassFromOptions(
+        Blockly.registry.Type.FLYOUTS_HORIZONTAL_TOOLBOX, workspace.options);
+  } else {
+    FlyoutClass = Blockly.registry.getClassFromOptions(
+        Blockly.registry.Type.FLYOUTS_VERTICAL_TOOLBOX, workspace.options);
+  }
+
+  if (!FlyoutClass) {
+    throw Error('Blockly.VerticalFlyout, Blockly.HorizontalFlyout or your own' +
+        ' custom flyout must be required.');
+  }
+  return new FlyoutClass(workspaceOptions);
+};
+
+/**
+ * Adds all the toolbox items to the toolbox.
+ * @param {!Array<!Blockly.utils.toolbox.ToolboxItemDef>} toolboxDef Array
+ *     holding objects containing information on the contents of the toolbox.
+ * @protected
+ */
+Blockly.Toolbox.prototype.addContents_ = function(toolboxDef) {
```

Comment:
`renderContents_` instead? As this is part of the render loop?

Reply:
done.